OpenVDB and Alembic reader

sosh opened the pull request:
Adding OpenVDB point reader contributed by Roman Zulak and Alembic point reader contributed by Lucas Miller.

lgritz commented:
Ping

davvid commented:
This is super cool. I'll file some internal issues so that we can have our partio maintainer chime in as well. I'll leave a few review notes here so we can push this forward.

davvid commented on src/lib/CMakeLists.txt:
```diff
@@ -46,6 +46,19 @@ set_target_properties(partio PROPERTIES
+set(TBB_INCLUDE_DIRS "$ENV{TBB_ROOT}/include")
+set(CMAKE_MODULE_PATH "$ENV{OpenVDB_ROOT}/lib/cmake/OpenVDB"
+    "${OpenVDB_ROOT}/lib/cmake/OpenVDB"
+    ${CMAKE_MODULE_PATH})
```
`CMAKE_MODULE_PATH` usually shouldn't need to be modified. If you set `CMAKE_PREFIX_PATH` to include the `OpenVDB_ROOT` then in theory `find_package(OpenVDB)` should work without needing to special-case openvdb here.

davvid commented on src/lib/CMakeLists.txt:
```diff
@@ -46,6 +46,19 @@ set_target_properties(partio PROPERTIES
+
+target_link_libraries(partio PRIVATE Alembic::Alembic OpenVDB::openvdb)
+target_include_directories(partio PRIVATE ${TBB_INCLUDE_DIRS})
+
```
Right now partio has a really minimal set of dependencies.

Since these are new dependencies it would be good to ease into them. Can we please make these opt-in and not built by default?

I would suggest adding some `option(...)` variables that default to `OFF`. Maybe call them something like `PARTIO_ENABLE_ALEMBIC` and `PARTIO_ENABLE_OPENVDB` so that we can opt-in to these settings.

TBB is needed for openvb so we can bundle those settings alongside `PARTIO_ENABLE_OPENVDB`.

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+{
+    // partio looks for the extension .abc to decide to call readABC
+    // and we want to be able to accept a time to read (right now in seconds)
+    // so we will pass in /tmp/foo@42.5.abc
```
Oh, bummer. That's really unfortunate. Would it help at all if we let you pass in something that looked like a URL? eg. `foo.abc?time=42.5`. Just thinking out loud for the future..

If we start to support this construct now then we will have to forever support it, so we should think about it.

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+            name = name.substr(0, lastAt) + ".abc";
+        }
+        catch(...)
+        {}
```
We should be explicit about what we're catching here. It looks like the idea is that if it's not something that converts via `stod` then we'll ignore the exception it throws?

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+            *errorStream << "Partio: Unable to open file Alembic: " << name << std::endl;
+            *errorStream << "(" << filename << ")" << std::endl;
+        }
+        return 0;
```
sug: prefer `nullptr` over `0`

davvid commented on src/lib/io/ParticleIO.cpp:
```diff
@@ -54,6 +54,8 @@ readers()
     static bool initialized=false;
     if(!initialized){
 	initializationMutex.lock();
```
if we make these plugins opt-in then we'll probably need `#defines` that match them so that we can conditionally compile these lines below.

davvid commented on src/lib/io/VDB.cpp:
```diff
@@ -0,0 +1,371 @@
+    }
+    tbb::task_scheduler_init schedulerInit((numThreads == 0) ? tbb::task_scheduler_init::automatic : numThreads);
+
+    initLibrary();
```
initLibrary() and all this TBB scheduler initialization stuff seems like it's `main()`'s responsibility rather than the responsibility of a plugin.

`CUE_THREADS` is rather specific and seeing `openvb::uninitialize()` makes me a bit nervous because I'm assuming that it's touching some global state.

I think this would be better mentioned in the readme with some example code showing what users need to do in order to use the library rather than taking on these responsibilities inside library code.

Try to keep these parts generic and reusable with as little assumptions as possible.

davvid commented on src/lib/io/VDB.cpp:
```diff
@@ -0,0 +1,371 @@
+    struct AttributeCache {
+        std::vector<AttributeIndex> indices;
+        ParticleAttribute           partio;    // partio.type contains the type that partio will store / allocate
+        unsigned                    tupleSize;
```
I dunno if it's just me, but I prefer seeing `unsigned int` over `unsigned` even though I know it's implied.

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+                            oAttrMap[keyName] = ioData->addAttribute(
+                                propName.c_str(), INT, extent);
+                        }
+                    }
```
does this need `break;` to avoid fallthrough?

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+                            oAttrMap[keyName] = ioData->addAttribute(
+                                propName.c_str(), ptype, extent);
+                        }
+                    }
```
ditto, do we need break above and below?

davvid commented on src/lib/io/readers.h:
```diff
@@ -36,6 +36,8 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 namespace Partio{
+ParticlesDataMutable* readABC(  const char* filename,const bool headersOnly,std::ostream* errorStream);
+ParticlesDataMutable* readVDB(  const char* filename,const bool headersOnly,std::ostream* errorStream);
 ParticlesDataMutable* readBGEO(	const char* filename,const bool headersOnly,std::ostream* errorStream);
```
note: this part would also nee some special ifdef handling if we made openvb + alembic optional.

davvid commented on src/lib/io/ABC.cpp:
```diff
@@ -0,0 +1,429 @@
+
+        std::map<std::string, ParticleAttribute>::const_iterator it;
+        for (it = iAttrMap.begin(); it != iAttrMap.end(); ++it)
+        {
```
Tiny style sug ~ map iterators are one of those areas where `auto` is kind of nice. Could we even go one step further and use a new-style loop here?

```
for (const auto& entry : iAttrMap) {
   // entry.first, entry.second
}
```

davvid reviewed (changes requested):
This looks great, thanks again for getting this working.

We'll see if others have any notes but this looks like a good place to start refining. I'm not opposed to merging this in if this feature can be made optional / opt-in and the implementation issues are be resolved.

lgritz commented:
@sosh @davvid Somebody asked me about this recently, I had forgotten that it never quite got merged, so I'm poking at it in the hope of reviving this zombie. 

Can we all somehow address the suggestions, put a bow on this, and get it merged? I think others would like the functionality.

sosh commented:
Larry, I believe Lucas Miller already addressed most of the comment from
David in our internal git at SPI.(01/04/2023)

On Mon, Jan 20, 2025 at 10:44 AM Larry Gritz ***@***.***>
wrote:

> @sosh <https://github.com/sosh> @davvid <https://github.com/davvid>
> Somebody asked me about this recently, I had forgotten that it never quite
> got merged, so I'm poking at it in the hope of reviving this zombie.
>
> Can we all somehow address the suggestions, put a bow on this, and get it
> merged? I think others would like the functionality.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wdas/partio/pull/102#issuecomment-2603060867>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADV42DHINA4BG2OYYK34OL2LU7XNAVCNFSM6AAAAABVQ6NCO2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBTGA3DAOBWG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lgritz commented:
Can you please push that update to GitHub so it can get reviewed and merged?

davvid commented:
Thanks for the pings. I'd be more than happy to merge this as soon as the new code is pushed.

In case it helps ~ feel free to ignore my comments about making this feature opt-in. I can wrangle that topic post-merge by updating the cmake scripts and adding a few conditional ifdefs to the code.

Thanks for making this happen, much appreciated.
